server/src/queries.rs

```diff
@@ -0,0 +1,252 @@
+use crate::router_utils::{
+    estimate_flight_time_minutes, get_nearby_nodes, get_route, init_router, Aircraft,
+    NearbyLocationQuery, RouteQuery, LANDING_AND_UNLOADING_TIME_MIN, LOADING_AND_TAKEOFF_TIME_MIN,
+    NODES, SAN_FRANCISCO,
+};
+use crate::scheduler_grpc::{
+    CancelFlightResponse, ConfirmFlightResponse, FlightPriority, FlightStatus, Id, QueryFlightPlan,
+    QueryFlightRequest, QueryFlightResponse,
+};
+use std::collections::HashMap;
+
+use crate::calendar_utils::{Calendar, Tz};
+use chrono::{Duration, NaiveDateTime, TimeZone};
+use once_cell::sync::OnceCell;
+use prost_types::Timestamp;
+use std::str::FromStr;
+use std::sync::Mutex;
+use std::time::SystemTime;
+use svc_storage_client_grpc::client::{
+    storage_rpc_client::StorageRpcClient, AircraftFilter, FlightPlan, VertiportFilter,
+};
+use tokio;
+use tonic::{Request, Response, Status};
+use uuid::Uuid;
+
+const CANCEL_FLIGHT_SECONDS: u64 = 30;
+
+fn unconfirmed_flight_plans() -> &'static Mutex<HashMap<String, FlightPlan>> {
+    static INSTANCE: OnceCell<Mutex<HashMap<String, FlightPlan>>> = OnceCell::new();
+    INSTANCE.get_or_init(|| Mutex::new(HashMap::new()))
+}
+
+fn cancel_flight_after_timeout(id: String) {
+    tokio::spawn(async move {
```

Comment:
I suppose this is one way to do it

I would maybe prefer a single thread that, at a known periodicity, retrieves all flight plans by a filter (DRAFT + Older than 30 seconds) and does cleanup rather than spinning a new thread for each query. While we're likely not going to see a billion threads be spun off for a billion queries in nominal operation, I could see this as a possible DDOS vulnerability

@owlot thoughts?

Reply:
In general I would like to avoid premature optimizations. Every tonic/tokio request spawns a thread anyway, so having 2x threads should not be a problem (it's not 10x or some exponential growth).
My preference is to remove caching completely and rely on storage/db as a single source of truth - e.g. create drafts there, then update status do confirmed. I never saw database clogged by simple updates and queries by ID, it's usually nested queries for analytics or huge insert/update batches which consume resources.

Reply:
Fair enough, thank you for the explanation